chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -951,4 +956,53 @@ impl<'a, S: BlockchainStorageWrite, O: OrphanBlocksMut, V: TransactionVerificati
             Err(err) => (*err).into(),
         }
     }
+
+    fn activate_seal_epoch(&mut self, height: BlockHeight) -> Result<(), BlockError> {
+        if is_due_for_epoch_data_calculation(self.chain_config, height) {
+            let current_epoch_index = self.chain_config.epoch_index_from_height(&height);
+            let sealed_epoch_distance_from_tip =
+                self.chain_config.sealed_epoch_distance_from_tip() as u64;
+
+            if current_epoch_index >= sealed_epoch_distance_from_tip {
+                let epoch_index_to_seal = current_epoch_index - sealed_epoch_distance_from_tip;
+
+                let first_block_epoch_to_seal =
+                    epoch_index_to_seal * self.chain_config.epoch_length().get();
+                let last_block_epoch_to_seal =
+                    first_block_epoch_to_seal + self.chain_config.epoch_length().get() - 1;
+
+                // iterate over every block in the epoch and merge every block delta into a singe delta
+                let epoch_delta = (first_block_epoch_to_seal..=last_block_epoch_to_seal)
```

Comment:
What about the undos?

Reply:
I don't store undos separately but reuse undos for regular transactions from `DBAccountingBlockUndo` index